jrt/src/com/yahoo/jrt/TransportThread.java

```diff
@@ -0,0 +1,352 @@
+        parent.handleFailure(problem, context);
+    }
+
+    /**
+     * Add a connection to the set of connections handled by this
+     * TransportThread. Invoked by the {@link Connector} class.
+     *
+     * @param conn the connection to add
+     **/
+    void addConnection(Connection conn) {
+        if (!postCommand(new AddConnectionCmd(conn))) {
+            perform(new CloseConnectionCmd(conn));
+        }
+    }
+
+    /**
+     * Request an asynchronous close of a connection.
+     *
+     * @param conn the connection to close
+     **/
+    void closeConnection(Connection conn) {
+        postCommand(new CloseConnectionCmd(conn));
+    }
+
+    /**
+     * Request an asynchronous enabling of write events for a
+     * connection.
+     *
+     * @param conn the connection to enable write events for
+     **/
+    void enableWrite(Connection conn) {
+        if (Thread.currentThread() == thread) {
+            handleEnableWrite(conn);
+        } else {
+            postCommand(new EnableWriteCmd(conn));
+        }
+    }
+
+    void handshakeWorkDone(Connection conn) {
+        postCommand(new HandshakeWorkDoneCmd(conn));
+    }
+
+    /**
+     * Create a {@link Task} that can be scheduled for execution in
+     * the transport thread.
+     *
+     * @return the newly created Task
+     * @param cmd what to run when the task is executed
+     **/
+    public Task createTask(Runnable cmd) {
+        return new Task(scheduler, cmd);
+    }
+
+    /**
+     * Perform the given command in such a way that it does not run
+     * concurrently with the transport thread or other commands
+     * performed by invoking this method. This method will continue to
+     * work even after the transport thread has been shut down.
+     *
+     * @param cmd the command to perform
+     **/
+    public void perform(Runnable cmd) {
+        if (Thread.currentThread() == thread) {
+            cmd.run();
+            return;
+        }
+        if (!postCommand(cmd)) {
+            join();
+            synchronized (thread) {
+                cmd.run();
+            }
+        }
+    }
+
+    /**
+     * Synchronize with the transport thread. This method will block
+     * until all commands issued before this method was invoked has
+     * completed. If the transport thread has been shut down (or is in
+     * the progress of being shut down) this method will instead wait
+     * for the transport thread to complete, since no more commands
+     * will be performed, and waiting would be forever. Invoking this
+     * method from the transport thread is not a good idea.
+     *
+     * @return this object, to enable chaining
+     **/
+    public TransportThread sync() {
+        SyncCmd cmd = new SyncCmd();
+        if (postCommand(cmd)) {
+            cmd.waitDone();
+        } else {
+            join();
+        }
+        return this;
+    }
+
+    private void run() {
+        while (state == OPEN) {
```

Comment:
This is not synchronized.
I suggest using an AtomicInt as locking does not seem necessary.

Reply:
now done in transport thread